<a href="https://colab.research.google.com/github/aparna-2001/machine_learning_practice/blob/main/level1_ML_foundation_BCI.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Alphawaves data,

subject 0

In [34]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy.io import loadmat


- numpy for numerical computing
- pandas for data handling
- matplotlib for plotting and visualising
- seaborn for visualising
- scipy, loadmat for loading the .mat files


In [35]:
#loading the .mat file in loadmat

data = loadmat('subject_00.mat')
print(data.keys())

print(type(data['SIGNAL']))

signal = data['SIGNAL']

if hasattr(signal, 'dtype') and signal.dtype.names:
    print(signal.dtype.names)


dict_keys(['__header__', '__version__', '__globals__', 'SIGNAL'])
<class 'numpy.ndarray'>


In [36]:
#print(data['__header__'])
#print(data['__version__'])
#print(data['__globals__'])
print(data['SIGNAL'])

[[ 0.00000000e+00  2.31471094e+03 -2.47386426e+03 ... -3.27888818e+03
   0.00000000e+00  0.00000000e+00]
 [ 1.95312500e-03  2.31207739e+03 -2.47719287e+03 ... -3.27667456e+03
   0.00000000e+00  0.00000000e+00]
 [ 3.90625000e-03  2.30813916e+03 -2.47775000e+03 ... -3.26313354e+03
   0.00000000e+00  0.00000000e+00]
 ...
 [ 1.23244141e+02  1.92760596e+03 -2.45507520e+03 ... -3.27909985e+03
   0.00000000e+00  0.00000000e+00]
 [ 1.23246094e+02  1.93911743e+03 -2.44151099e+03 ... -3.25797363e+03
   0.00000000e+00  0.00000000e+00]
 [ 1.23248047e+02  1.94014917e+03 -2.43970483e+03 ... -3.24809204e+03
   0.00000000e+00  0.00000000e+00]]


In [37]:
#extracting the signal only

signal_data = pd.DataFrame(data['SIGNAL'])

print(signal_data.shape)

signal_data.head(5)

(63104, 19)


0            1            2           3            4            5   \
0  0.000000  2314.710938 -2473.864258  304.726593 -6740.526367 -6507.790039   
1  0.001953  2312.077393 -2477.192871  295.271210 -6747.804199 -6508.391602   
2  0.003906  2308.139160 -2477.750000  292.862305 -6750.519043 -6510.719238   
3  0.005859  2303.175293 -2477.378174  296.507050 -6756.901856 -6515.480957   
4  0.007812  2307.331543 -2471.294434  301.831573 -6754.847656 -6509.622070   

             6            7             8            9            10  \
0 -11247.854492 -2859.524170 -14004.568359 -1706.199707 -4104.730469   
1 -11249.747070 -2863.340820 -14008.770508 -1708.395874 -4104.203613   
2 -11248.869141 -2867.504883 -14007.042969 -1709.802368 -4105.119141   
3 -11250.728516 -2872.019775 -14011.469727 -1709.260498 -4109.889648   
4 -11247.145508 -2865.405518 -14011.854492 -1702.394165 -4102.324707   

            11            12           13          14          15  \
0 -6290.920410 -12415.869141 -7290.736816  170.436508  397.110260   
1 -6288.968262 -12418.380859 -7293.556641  175.323837  400.196503   
2 -6274.008301 -12422.494141 -7294.418945  193.758057  410.232483   
3 -6266.696777 -12422.033203 -7302.818848  197.839844  410.054993   
4 -6257.812012 -12413.977539 -7298.030273  200.905090  415.010742   

            16   17   18  
0 -3278.888184  0.0  0.0  
1 -3276.674561  0.0  0.0  
2 -3263.133545  0.0  0.0  
3 -3260.899902  0.0  0.0  
4 -3255.357422  0.0  0.0

In [38]:
#signal_data.iloc[0:6, 0]

first_row = signal_data.iloc[0, 0]
second_row = signal_data.iloc[1, 0]
third_row = signal_data.iloc[2, 0]

print(second_row - first_row)
print(third_row - second_row)


0.001953125
0.001953125


In [39]:
#sampling rate

sampling_rate = 1/(signal_data.iloc[1,0] - signal_data.iloc[0, 0])

print(sampling_rate)

512.0


In [40]:
#recording time

recording_time = signal_data.iloc[-1,0] - signal_data.iloc[0,0]
print(recording_time)

123.24804688


In [41]:
#confirming the sample rate

sampling_rate_conf = signal_data.shape[0]/recording_time
print(sampling_rate_conf)

512.0081136980691


since the difference is uniform, the first column will be the time

- file is in .mat format.

- The data is structured as rows contain observation at each time.  

- sample and columns contains the recordings of EEG electrodes.

- sampling rate - total number of samples/ total recording time in seconds; or 1/(time[1] - time[0]) , sampling rate of this data is 512 Hz

- How long is the recording ; The recording time is 123.24804688 seconds, approximately close to 2 minutes


In [42]:
#shape of the data

print(signal_data.shape)

print(signal_data.columns,'\n')

print(signal_data.info())

(63104, 19)
RangeIndex(start=0, stop=19, step=1) 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 63104 entries, 0 to 63103
Data columns (total 19 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       63104 non-null  float64
 1   1       63104 non-null  float64
 2   2       63104 non-null  float64
 3   3       63104 non-null  float64
 4   4       63104 non-null  float64
 5   5       63104 non-null  float64
 6   6       63104 non-null  float64
 7   7       63104 non-null  float64
 8   8       63104 non-null  float64
 9   9       63104 non-null  float64
 10  10      63104 non-null  float64
 11  11      63104 non-null  float64
 12  12      63104 non-null  float64
 13  13      63104 non-null  float64
 14  14      63104 non-null  float64
 15  15      63104 non-null  float64
 16  16      63104 non-null  float64
 17  17      63104 non-null  float64
 18  18      63104 non-null  float64
dtypes: float64(19)
memory usage: 9.1 MB
None


Answer these questions by exploring:

  
*  The shape of the data is 63104 * 19
*  There are no missing values
*  Each data type is float


In [43]:
#summary statistics

signal_data_stat = signal_data.iloc[:, 1:17].describe()
signal_data_stat

1             2             3             4             5   \
count  63104.000000  63104.000000  63104.000000  63104.000000  63104.000000   
mean    2289.039524  -2341.967507    486.993320  -6781.180625  -6290.831646   
std      141.354372    111.627399     94.237564     91.141335    100.112292   
min     1919.117920  -2557.157959    217.782211  -7108.479492  -6534.169434   
25%     2181.048035  -2456.960937    430.085991  -6816.313232  -6374.752075   
50%     2275.398926  -2362.735352    510.669525  -6781.700684  -6266.024902   
75%     2422.695740  -2231.222046    560.539948  -6714.244019  -6208.651123   
max     2577.585205  -2104.002686    795.885376  -6612.576660  -6095.611328   

                 6             7             8             9             10  \
count  63104.000000  63104.000000  63104.000000  63104.000000  63104.000000   
mean  -11416.863099  -2595.009038 -13955.873626  -1795.112935  -3777.377751   
std      113.830209    130.498954     38.852913     74.720035    156.177567   
min   -11849.524414  -2934.821289 -14278.633789  -2057.885986  -4158.770019   
25%   -11518.110351  -2712.221619 -13982.040039  -1844.913513  -3908.614929   
50%   -11400.988769  -2555.023926 -13955.098633  -1780.144714  -3739.749512   
75%   -11323.991700  -2486.022583 -13930.370361  -1737.875031  -3644.156067   
max   -11208.399414  -2391.815186 -13821.010742  -1659.822144  -3515.339600   

                 11            12            13            14            15  \
count  63104.000000  63104.000000  63104.000000  63104.000000  63104.000000   
mean   -6001.114902 -12319.253217  -7259.949999     61.228166    440.071618   
std      161.860930     48.089418     34.012791     49.931331     47.053957   
min    -6366.761231 -12492.496094  -7381.578613    -59.928364    303.357452   
25%    -6145.006348 -12344.897461  -7282.695435     26.797468    401.499443   
50%    -5971.622803 -12311.807129  -7263.285156     52.249720    438.953949   
75%    -5855.889893 -12282.636963  -7240.133667     84.224888    475.364578   
max    -5726.430176 -12200.150391  -7144.030273    241.197327    592.525391   

                 16  
count  63104.000000  
mean   -3237.453961  
std       32.626487  
min    -3355.604248  
25%    -3259.574646  
50%    -3237.737305  
75%    -3215.229431  
max    -3124.399170

In [47]:
signal_data_stat_transpose = signal_data_stat.transpose()
signal_data_stat_transpose


count          mean         std           min           25%  \
1   63104.0   2289.039524  141.354372   1919.117920   2181.048035   
2   63104.0  -2341.967507  111.627399  -2557.157959  -2456.960937   
3   63104.0    486.993320   94.237564    217.782211    430.085991   
4   63104.0  -6781.180625   91.141335  -7108.479492  -6816.313232   
5   63104.0  -6290.831646  100.112292  -6534.169434  -6374.752075   
6   63104.0 -11416.863099  113.830209 -11849.524414 -11518.110351   
7   63104.0  -2595.009038  130.498954  -2934.821289  -2712.221619   
8   63104.0 -13955.873626   38.852913 -14278.633789 -13982.040039   
9   63104.0  -1795.112935   74.720035  -2057.885986  -1844.913513   
10  63104.0  -3777.377751  156.177567  -4158.770019  -3908.614929   
11  63104.0  -6001.114902  161.860930  -6366.761231  -6145.006348   
12  63104.0 -12319.253217   48.089418 -12492.496094 -12344.897461   
13  63104.0  -7259.949999   34.012791  -7381.578613  -7282.695435   
14  63104.0     61.228166   49.931331    -59.928364     26.797468   
15  63104.0    440.071618   47.053957    303.357452    401.499443   
16  63104.0  -3237.453961   32.626487  -3355.604248  -3259.574646   

             50%           75%           max  
1    2275.398926   2422.695740   2577.585205  
2   -2362.735352  -2231.222046  -2104.002686  
3     510.669525    560.539948    795.885376  
4   -6781.700684  -6714.244019  -6612.576660  
5   -6266.024902  -6208.651123  -6095.611328  
6  -11400.988769 -11323.991700 -11208.399414  
7   -2555.023926  -2486.022583  -2391.815186  
8  -13955.098633 -13930.370361 -13821.010742  
9   -1780.144714  -1737.875031  -1659.822144  
10  -3739.749512  -3644.156067  -3515.339600  
11  -5971.622803  -5855.889893  -5726.430176  
12 -12311.807129 -12282.636963 -12200.150391  
13  -7263.285156  -7240.133667  -7144.030273  
14     52.249720     84.224888    241.197327  
15    438.953949    475.364578    592.525391  
16  -3237.737305  -3215.229431  -3124.399170

In [46]:
headers = ['FP1', 'FP2', 'FC5', 'FC6', 'FZ', 'T7', 'CZ', 'T8',
           'P7', 'P3', 'PZ', 'P4', 'P8', 'O1', 'Oz', 'O2']

# Create a dictionary to map old column names (integers 1-16) to new headers
column_mapping = {i + 1: header for i, header in enumerate(headers)}

#renaming the columns
signal_data_stat_set = signal_data_stat.rename(columns=column_mapping)

signal_data_stat_set

FP1           FP2           FC5           FC6            FZ  \
count  63104.000000  63104.000000  63104.000000  63104.000000  63104.000000   
mean    2289.039524  -2341.967507    486.993320  -6781.180625  -6290.831646   
std      141.354372    111.627399     94.237564     91.141335    100.112292   
min     1919.117920  -2557.157959    217.782211  -7108.479492  -6534.169434   
25%     2181.048035  -2456.960937    430.085991  -6816.313232  -6374.752075   
50%     2275.398926  -2362.735352    510.669525  -6781.700684  -6266.024902   
75%     2422.695740  -2231.222046    560.539948  -6714.244019  -6208.651123   
max     2577.585205  -2104.002686    795.885376  -6612.576660  -6095.611328   

                 T7            CZ            T8            P7            P3  \
count  63104.000000  63104.000000  63104.000000  63104.000000  63104.000000   
mean  -11416.863099  -2595.009038 -13955.873626  -1795.112935  -3777.377751   
std      113.830209    130.498954     38.852913     74.720035    156.177567   
min   -11849.524414  -2934.821289 -14278.633789  -2057.885986  -4158.770019   
25%   -11518.110351  -2712.221619 -13982.040039  -1844.913513  -3908.614929   
50%   -11400.988769  -2555.023926 -13955.098633  -1780.144714  -3739.749512   
75%   -11323.991700  -2486.022583 -13930.370361  -1737.875031  -3644.156067   
max   -11208.399414  -2391.815186 -13821.010742  -1659.822144  -3515.339600   

                 PZ            P4            P8            O1            Oz  \
count  63104.000000  63104.000000  63104.000000  63104.000000  63104.000000   
mean   -6001.114902 -12319.253217  -7259.949999     61.228166    440.071618   
std      161.860930     48.089418     34.012791     49.931331     47.053957   
min    -6366.761231 -12492.496094  -7381.578613    -59.928364    303.357452   
25%    -6145.006348 -12344.897461  -7282.695435     26.797468    401.499443   
50%    -5971.622803 -12311.807129  -7263.285156     52.249720    438.953949   
75%    -5855.889893 -12282.636963  -7240.133667     84.224888    475.364578   
max    -5726.430176 -12200.150391  -7144.030273    241.197327    592.525391   

                 O2  
count  63104.000000  
mean   -3237.453961  
std       32.626487  
min    -3355.604248  
25%    -3259.574646  
50%    -3237.737305  
75%    -3215.229431  
max    -3124.399170



## **EEG Channel Summary Statistics (Columns 1-16 Only)**

| Channel | Electrode | Mean | Std | Min | Max | Range | Baseline Status | Signal Quality |
|---------|-----------|----------|---------|-------------|-------------|---------|-----------------|----------------|
| 1 | **FP1** | 2289.04 | 141.35 | 1919.12 | 2577.59 | 658.47 | Moderate Drift | High Variability |
| 2 | **FP2** | -2341.97 | 111.63 | -2557.16 | -2104.00 | 453.16 | Moderate Drift | Normal |
| 3 | **FC5** | 486.99 | 94.24 | 217.78 | 795.89 | 578.11 | ✓ Good | Normal |
| 4 | **FC6** | -6781.18 | 91.14 | -7108.48 | -6612.58 | 495.90 | ⚠️ Severe Drift | Normal |
| 5 | **FZ** | -6290.83 | 100.11 | -6534.17 | -6095.61 | 438.56 | ⚠️ Severe Drift | Normal |
| 6 | **T7** | -11416.86 | 113.83 | -11849.52 | -11208.40 | 641.12 | ⚠️ Severe Drift | Normal |
| 7 | **CZ** | -2595.01 | 130.50 | -2934.82 | -2391.82 | 543.00 | Moderate Drift | High Variability |
| 8 | **T8** | -13955.87 | 38.85 | -14278.63 | -13821.01 | 457.62 | ⚠️ **CRITICAL** | **Flatline Risk** |
| 9 | **P7** | -1795.11 | 74.72 | -2057.89 | -1659.82 | 398.07 | Moderate Drift | Normal |
| 10 | **P3** | -3777.38 | 156.18 | -4158.77 | -3515.34 | 643.43 | Moderate Drift | High Variability |
| 11 | **PZ** | -6001.11 | 161.86 | -6366.76 | -5726.43 | 640.33 | ⚠️ Severe Drift | High Variability |
| 12 | **P4** | -12319.25 | 48.09 | -12492.50 | -12200.15 | 292.35 | ⚠️ **CRITICAL** | **Flatline Risk** |
| 13 | **P8** | -7259.95 | 34.01 | -7381.58 | -7144.03 | 237.55 | ⚠️ **CRITICAL** | **Flatline Risk** |
| 14 | **O1** | 61.23 | 49.93 | -59.93 | 241.20 | 301.13 | ✓ Good | Low Variability |
| 15 | **Oz** | 440.07 | 47.05 | 303.36 | 592.53 | 289.17 | ✓ Good | Low Variability |
| 16 | **O2** | -3237.45 | 32.63 | -3355.60 | -3124.40 | 231.20 | Moderate Drift | Low Variability |

---

## **Summary by Brain Region:**

| Region | Channels | Good Baseline | Moderate Drift | Severe Drift | Critical Issues |
|--------|----------|---------------|----------------|--------------|-----------------|
| **Frontal** | FP1, FP2, FC5, FC6, FZ | FC5 ✓ | FP1, FP2 | FC6, FZ | None |
| **Temporal** | T7, T8 | None | None | T7 | **T8 (Flatline)** |
| **Central** | CZ | None | CZ | None | None |
| **Parietal** | P7, P3, PZ, P4, P8 | None | P7, P3 | PZ | **P4, P8 (Flatline)** |
| **Occipital** | O1, Oz, O2 | O1, Oz ✓ | O2 | None | None |



In [53]:
# which channel has the highest variability

print(signal_data_stat_transpose['std'].min())
print(signal_data_stat_transpose['std'].max())

#channel 11, pz

32.62648744905857
161.86093012937283


In [54]:
#Which channel has the largest amplitude?

print(signal_data_stat_transpose['mean'].min())
print(signal_data_stat_transpose['mean'].max())

#channle 8, T8

-13955.873625734246
2289.0395237703583


In [ ]:
#Why might different channels behave differently?
